### PR TITLE
[vLLM] Support vLLM CPU backend and provide QBits acceleration

### PR DESCRIPTION
**Pull request by Zhenzhong1**
## Type of Change

New feature & API change

## Description

- [x] Complete the pipeline to accelerate vLLM linear modules by qbits.
- [x] API Modification: vllm_model = AutoModelForCausalLM.from_pretrained(args.model, use_vllm = True)
- [x] Supported pytorch==2.3.0 + cpu & vLLM==0.4.2 + cpu
- [x] Adapated to more vLLM modules

All models listed in the [vLLM Supported Models](https://docs.vllm.ai/en/latest/models/supported_models.html) can be accelerated theoretically.

Validated models:
* [THUDM/chatglm2-6b](https://hf-mirror.com/THUDM/chatglm2-6b)
* [meta-llama/Llama-2-7b-chat-hf](https://hf-mirror.com/meta-llama/Llama-2-7b-chat-hf)
* [baichuan-inc/Baichuan2-7B-Chat](https://hf-mirror.com/baichuan-inc/Baichuan2-7B-Chat)
* [tiiuae/falcon-7b](https://huggingface.co/tiiuae/falcon-7b)
* [meta-llama/Meta-Llama-3-8B](https://huggingface.co/meta-llama/Meta-Llama-3-8B)
* [mistralai/Mistral-7B-Instruct-v0.1](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1)
* [microsoft/phi-2](https://huggingface.co/microsoft/phi-2)
* [Qwen/CodeQwen1.5-7B-Chat](https://huggingface.co/Qwen/CodeQwen1.5-7B-Chat)

Performance:
```text
# Env: SPR numactl -m 0 -C 0-55
# Model: Llama2-7b-chat-hf

32 in, 32 out
The Neural Speed:  942 ms  (Generation Speed: 33.97 toks/s, 213ms first token * 1+ 23.5ms next token * 31)
The qbits    vLLM: 1677 ms (Generation Speed: 19.08 toks/s)
The original vLLM: 2304 ms (Generation Speed: 13.75 toks/s)

1024 in 32 out
The Neural Speed  : 2344ms  (Generation Speed: 13.65 toks/s, 1527ms first token * 1+ 31 next_token * 26.38ms)
The qbits    vLLM : 2912ms  (Generation Speed: 11.0 toks/s)
The original vLLM : 3575ms  (Generation Speed: 8.95 toks/s)
```

## Expected Behavior & Potential Risk

N/A

## How has this PR been tested?

Nightly unit tests.

Llama2-7b
image

## Dependency Change?

added in the example/vllm/requirement.txt

### CODE REVIEW

File: examples/vllm/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+```
+
+## Supported and Validated Models
+All models listed in the [vLLM Supported Models](https://docs.vllm.ai/en/latest/models/supported_models.html) can be accelerated theoretically.
+
+We have validated the majority of existing models using vLLM==0.4.2+cpu:
+* [THUDM/chatglm2-6b](https://hf-mirror.com/THUDM/chatglm2-6b)
+* [meta-llama/Llama-2-7b-chat-hf](https://hf-mirror.com/meta-llama/Llama-2-7b-chat-hf)
+* [baichuan-inc/Baichuan2-7B-Chat](https://hf-mirror.com/baichuan-inc/Baichuan2-7B-Chat)
+
```
Comment (a32543254) [inline examples/vllm/README.md:29]: how about also provide  the accuracy option for users
please refer to https://github.com/intel/intel-extension-for-transformers/tree/main/examples/huggingface/pytorch/text-generation/quantization

File: tests/Nightly/test_neural_speed.py [hunk 2]
```diff
@@ -95,6 +95,14 @@ def test_gguf_api(self):
         assert(output[0][0:15] == [1, 5713, 3714, 264, 727, 28725, 736, 403, 264, 1628, 2746, 693, 6045, 298, 1220])
 
 
+    def test_use_vllm_api(self):
+        model_name = "/tf_dataset2/models/nlp_toolkit/llama-2-7b-chat/Llama-2-7b-chat-hf"
+        prompt = "Once upon a time"
+        model = AutoModelForCausalLM.from_pretrained(model_name, use_vllm = True)
```
Comment (a32543254) [inline tests/Nightly/test_neural_speed.py:101]: better to add a new ut script, rather than use neural speed one.

**Review by a32543254 — APPROVED** [review]
LGTM

File: examples/vllm/vllm_acceleration_example.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import time
+import os
+from vllm import LLM, SamplingParams
+from typing import List, Optional
+from intel_extension_for_transformers.transformers import AutoModelForCausalLM, RtnConfig
+from transformers import AutoTokenizer
+
+
+def main(args_in: Optional[List[str]] = None) -> None:
+    parser = argparse.ArgumentParser(description="Convert a PyTorch model to a NE compatible file")
```
Comment (a32543254) [inline examples/vllm/vllm_acceleration_example.py:25]: Convert a PyTorch model to a NE compatible file
>>>>
Convert a PyTorch model to a NS compatible file

File: examples/vllm/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+numactl -m 0 -C 0-55 python vllm_acceleration_example.py --model_path=/home/model/chatglm2-6b --prompt=你好
+```
+
+## Supported and Validated Models
+All models listed in the [vLLM Supported Models](https://docs.vllm.ai/en/latest/models/supported_models.html) can be accelerated theoretically.
+
+We have validated the majority of existing models using vLLM==0.4.2+cpu:
+* [THUDM/chatglm2-6b](https://hf-mirror.com/THUDM/chatglm2-6b)
+* [meta-llama/Llama-2-7b-chat-hf](https://hf-mirror.com/meta-llama/Llama-2-7b-chat-hf)
+* [baichuan-inc/Baichuan2-7B-Chat](https://hf-mirror.com/baichuan-inc/Baichuan2-7B-Chat)
```
Comment (a32543254) [inline examples/vllm/README.md:28]: what about add them all here ? 
Validated models:

[THUDM/chatglm2-6b](https://hf-mirror.com/THUDM/chatglm2-6b)
[meta-llama/Llama-2-7b-chat-hf](https://hf-mirror.com/meta-llama/Llama-2-7b-chat-hf)
[baichuan-inc/Baichuan2-7B-Chat](https://hf-mirror.com/baichuan-inc/Baichuan2-7B-Chat)
[tiiuae/falcon-7b](https://huggingface.co/tiiuae/falcon-7b)
[meta-llama/Meta-Llama-3-8B](https://huggingface.co/meta-llama/Meta-Llama-3-8B)
[mistralai/Mistral-7B-Instruct-v0.1](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1)
[microsoft/phi-2](https://huggingface.co/microsoft/phi-2)

Comment (Zhenzhong1) [reply examples/vllm/README.md:28]: Yes, I will add them to the README later.

File: tests/Nightly/test_neural_speed.py (hunk 2 above)
Comment (Zhenzhong1) [reply tests/Nightly/test_neural_speed.py]: fixed

File: examples/vllm/vllm_acceleration_example.py (hunk 3 above)
Comment (Zhenzhong1) [reply examples/vllm/vllm_acceleration_example.py:25]: fixed

File: examples/vllm/README.md (hunk 1 above)
Comment (a32543254) [reply examples/vllm/README.md:34]: synced with zhenzhong, Will add acc in next pr.

Comment (Zhenzhong1) [reply examples/vllm/README.md:34]: I got this. I will add the accuracy mode in the next PR.

**Comment by kevinintel** [comment]
please provide perf vs pure vllm and ns

**Review by changwangss — APPROVED** [review]

**Comment by VincyZhang** [comment]
UT: https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/nlp-toolkit-optimize-ut/3148/

**Comment by VincyZhang** [comment]
https://inteltf-jenk.sh.intel.com/job/nlp-toolkit-optimize-ut/3149/

**Pull request merged**